chrome/content/zotero/xpcom/pdfWorker/manager.js

```diff
@@ -410,6 +410,148 @@ class PDFWorker {
 		}
 		await Promise.all(promises);
 	}
+
+	/**
+	 * Delete pages from PDF attachment
+	 *
+	 * @param {Integer} itemID Attachment item id
+	 * @param {Array} pageIndexes
+	 * @param {Boolean} [isPriority]
+	 * @param {String} [password]
+	 * @returns {Promise}
+	 */
+	async deletePages(itemID, pageIndexes, isPriority, password) {
+		return this._enqueue(async () => {
+			let attachment = await Zotero.Items.getAsync(itemID);
+
+			Zotero.debug(`Deleting [${pageIndexes.join(', ')}] pages for item ${attachment.libraryKey}`);
+			let t = new Date();
+
+			if (!attachment.isPDFAttachment()) {
+				throw new Error('Item must be a PDF attachment');
+			}
+
+			let annotations = attachment
+				.getAnnotations()
+				.map(annotation => ({
+					id: annotation.id,
+					position: JSON.parse(annotation.annotationPosition)
+				}));
+
+			let path = await attachment.getFilePathAsync();
+			let buf = await OS.File.read(path, {});
+			buf = new Uint8Array(buf).buffer;
+
+			try {
+				var { buf: modifiedBuf } = await this._query('deletePages', {
+					buf, pageIndexes, password
+				}, [buf]);
+			}
+			catch (e) {
+				let error = new Error(`Worker 'deletePages' failed: ${JSON.stringify({ error: e.message })}`);
+				try {
+					error.name = JSON.parse(e.message).name;
+				}
+				catch (e) {
+					Zotero.logError(e);
+				}
+				Zotero.logError(error);
+				throw error;
+			}
+
+			// Delete annotations from deleted pages
+			let ids = [];
+			for (let i = annotations.length - 1; i >= 0; i--) {
+				let { id, position } = annotations[i];
+				if (pageIndexes.includes(position.pageIndex)) {
+					ids.push(id);
+					annotations.splice(i, 1);
+				}
+			}
+			if (ids.length) {
+				await Zotero.Items.erase(ids);
+			}
+
+			// Shift page index for other annotations
+			let notifierQueue = new Zotero.Notifier.Queue();
```

Comment:
Did you try doing this in an actual DB transaction? For a bunch of items, it should be faster.

Reply:
Ok, I changed that, but you have to take a look to make sure it's properly implemented and safe.